### PR TITLE
TINY-3785: Expose TinyMCE public types and rename bridge public types from "api" to "spec"

### LINKED ISSUE #3427: TypeScript declaration file
Are there plans to add a TinyMCE declaration file for TypeScript? It is now [recommended](http://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html) that the file gets bundled with the npm package instead of getting added to [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped). I can provide a declaration file. It's still a WIP but could get things started.

### PR DESCRIPTION
**Pull request by lnewson**
Related Ticket: TINY-3785

Description of Changes:
This finally adds the ability to publish our types for TinyMCE. Since we now need to expose the bridge public types, I've done some cleanup there (added new `Public` modules and renamed `Api` to `Spec` to avoid confusion and renamed/cleaned up `Types` to avoid having a deeply nested tree). I've exposed the bridge types via `Ui.Dialog`, `Ui.Toolbar`, `Ui.Menu`, etc... and updated the plugins to use that instead of bridge directly. I've also gone through and cleaned up our public types to make sure they are correct and to remove a few more `any` types.

Pre-checks:
* [x] Changelog entry added
* [x] ~Tests have been added (if applicable)~
* [x] Branch prefixed with `feature/` for new features (if applicable)
* [x] ~License headers added on new files (if applicable)~
* [x] Swag dependency updated

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):
Fixes #3427

### CODE REVIEW

**Comment by lnewson** [comment]
I've marked this as ready for review, but it will continue to fail to build on CI until the swag version bump is done.

File: modules/tinymce/src/core/main/ts/api/util/EventDispatcher.ts [hunk 1]
```diff
@@ -8,15 +8,6 @@
 import { Fun, Obj } from '@ephox/katamari';
 import Tools from './Tools';
 
-// InputEvent is experimental so we don't have an actual type
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/util/EventDispatcher.ts:11]: Removed this as the `InputEvent` type is in the TS dom types now.

File: modules/bridge/src/main/ts/ephox/bridge/api/InlineContent.ts [hunk 2]
```diff
@@ -1,27 +1,44 @@
...
-  AutocompleterItemApi,
-  AutocompleterItem,
-  AutocompleterContents,
-  SeparatorItemApi,
-  SeparatorItem,
-  AutocompleterInstanceApi
+  Autocompleter, AutocompleterContents, AutocompleterInstanceApi, AutocompleterItem, AutocompleterItemSpec, AutocompleterSpec, ColumnTypes,
+  createAutocompleter, createAutocompleterItem, createSeparatorItem, SeparatorItem, SeparatorItemSpec
 } from '../components/content/Autocompleter';
+import { ContextPosition, ContextScope } from '../components/content/ContextBar';
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/api/InlineContent.ts:5]: I moved these over from `Toolbar`, as they are inline content UI components similar to the autocompleter. I also splint `ContextToolbar` into `ContextForm` and `ContextToolbar`.

File: modules/tinymce/src/core/main/ts/api/PublicApi.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
+/**
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/PublicApi.ts:1]: This is the file used to define the public API. It's then compiled and used by the new rollup plugin to bundle the final .d.ts file.

Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/PublicApi.ts:1]: If anyone can think of something else that should be exposed here, let me know.

File: modules/tinymce/src/core/main/ts/api/Resource.ts [hunk 4]
```diff
@@ -9,8 +9,8 @@ import ScriptLoader from './dom/ScriptLoader';
 import Promise from './util/Promise';
 
 interface Resource {
-  load(id: string, url: string): Promise<any>;
-  add(id: string, data: any): void;
+  load <T = any>(id: string, url: string): Promise<T>;
+  add (id: string, data: any): void;
```
Comment (ltrouton) [inline modules/tinymce/src/core/main/ts/api/Resource.ts:13]: Was wondering if this should match up with the `add` function below?

Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/Resource.ts:13]: Oops, I originally changed both and then reverted it. I'll revert the one below 👍

**Pull request merged**